### DISCUSSION TITLE
How to use Supabase with SWR?

### DISCUSSION THREAD

**Discussion by roker15** [supabase/supabase#3145] [Questions] [answered]
I see something like this 
`const { data, error } = useSWR(`/api/user/${id}`, fetcher)`
Then what will be the `fetcher` in case of super base?

I tried following but did not worked 
```

const fetcher = async () =>  await supabase.from("papers").select("*");

 const { data } = useSWR("/api/article", fetcher);
```

**Comment by thorwebdev** [comment]
I haven't looked into it, but you might find https://github.com/EdgarBarrantes/supabase-utils useful.

**Comment by MildTomato** [comment]
it looks like a rather odd way of using SWR - it looks like you may have gotten that url "api/user/${id}" from the [SWR docs](https://swr.vercel.app/docs/getting-started#make-it-reusable), but it looks like you just want to run supabase-js in client.

I think if you really wanted to use SWR here, it might look a bit like this.
(not tested this though).

```js
const fetcher = async (table) =>  await supabase.from(table).select("*");

function App () {
  const { data, error } = useSWR('papers', fetcher)
 
  if (!data) return "loading"
  // ...
}
```

It think this is rather obtuse though, since if you are just getting the data client side, then you can just use supabase-js and drop SWR and it will behave pretty much the same.

**Reply by roker15** [reply]
> It think this is rather obtuse though, since if you are just getting the data client side, then you can just use supabase-js and drop SWR and it will behave pretty much the same.

SWR does client side caching --- supabase also does that?

**Reply by roker15** [reply]
> It think this is rather obtuse though, since if you are just getting the data client side, then you can just use supabase-js and drop SWR and it will behave pretty much the same.

SWR does client side caching --- supabase also does that?

**Reply by MildTomato** [reply]
I'm not familiar with SWR cache - I don't think I'd be much help on this

**Reply by marc-on-github** [reply]
I tried it out. A part of the supabase-js library works like swr. Per page the data is loaded only once. SWR has a revalidate function, the supabase-js library doesn't have that. So if I have two browser windows open and make changes in one, they appear when I click into the other browser window too.

**Reply by mhsnook** [reply]
**edit: I moved this reply to [a top-level comment here](https://github.com/supabase/supabase/discussions/3145#discussioncomment-1426310).**


_**tl;dr:** Sure, you can use supabase calls in fetcher functions for useSWR; as long as your fetcher returns a promise of data, or throws an error, useSWR will give you the DX and UX comforts of modern Hooks, can save you from having to use Suspense and ContextProviders, and help you write cleaner more declarative code._

Because supabase-js client queries return a promise just like a fetcher function, you can use SWR to wrap these calls to return a react Hook and act as a sort of promise-and-error boundary. Actually, while supabase does seem to be doing some caching, it is structurally a lot like a replacement for a simple `fetch`. 

Wrapping these asynchronous calls in the useSWR hook means you don't have to worry about suspense or useEffect, and you get to write declarative components that know how they should present when their data is loading / loaded / error, and that update automatically across all tabs and all components when the promise returns. 

I'm still using code very similar to this useUser example from the docs, but I changed mine to useSession:

```javascript
export function useSession() {
  const { data: session, error } = useSWR('check_login', () => supabase.auth.session())
  return error ? { error } : {
    session,
    user: session?.user,
    isLoading: !error && !session,
  }
}
```

And then on a page where I want to check authentication and then load data, I have some code like this:

```javascript
  const { session } = useSession()
  const { data: posts, error } = useSWR(
    session ? 'posts/drafts' : null,
    getDraftPosts
  )
```

You can see I'm using the same cache keys as before I switched to Supabase (e.g. `posts/drafts`), which works fine for this situation even though the fetcher function doesn't use it, because it just acts as a cache-key that makes sense to humans reading the code like me.

And finally the fetcher function looks like this:

```javascript
export async function getDraftPosts() {
  const { data: posts, error } = await supabase
    .from('posts')
    .select('*')
    .eq('published', false)

  if (error) throw error.message
  return  posts
}
```

I haven't run into much weirdness yet and things are working fine. I will say that `auth.user()` and `auth.session()` just return the requested data rather than the `{ data, error }` returned by the others. So you can see I wrote a very thin fetcher function to get draft posts, destructure its return values, throw any errors and return the data. Whereas with `useSession` I just pass an anonymous arrow function as the fetcher: `useSWR('check_login', () => supabase.auth.session())` because `auth.session()` is _only_ returning the session object itself, or it's throwing an error; no errors-in-data and no destructuring needed.

Anyway I hope this helps a bit and good luck on your project!

**Comment by thorwebdev** [comment]
@michaelsnook thanks so much for your detailed writeup above! 💚  

Linking it here so I can mark it as the answer: https://github.com/supabase/supabase/discussions/3145#discussioncomment-1422090

All credit to you though!

**Reply by mhsnook** [reply]
ooh okay I just moved my reply to a top-level comment [here](https://github.com/supabase/supabase/discussions/3145#discussioncomment-1426310)

**Comment by mhsnook** [comment]
_**tl;dr:** Sure, you can use supabase calls in fetcher functions for useSWR; as long as your fetcher returns a promise of data, or throws an error, useSWR will give you the DX and UX comforts of modern Hooks, can save you from having to use Suspense and ContextProviders, and help you write cleaner more declarative code._

Because supabase-js client queries return a promise just like a fetcher function, you can use SWR to wrap these calls to return a react Hook and act as a sort of promise-and-error boundary. Actually, while supabase does seem to be doing some caching, it is structurally a lot like a replacement for a simple `fetch`. 

Wrapping these asynchronous calls in the useSWR hook means you don't have to worry about suspense or useEffect, and you get to write declarative components that know how they should present when their data is loading / loaded / error, and that update automatically across all tabs and all components when the promise returns. 

I'm still using code very similar to this useUser example from the docs, but I changed mine to useSession:

```javascript
export function useSession() {
  const { data: session, error } = useSWR('check_login', () => supabase.auth.session())
  return error ? { error } : {
    session,
    user: session?.user,
    isLoading: !error && !session,
  }
}
```

And then on a page where I want to check authentication and then load data, I have some code like this:

```javascript
  const { session } = useSession()
  const { data: posts, error } = useSWR(
    session ? 'posts/drafts' : null,
    getDraftPosts
  )
```

You can see I'm using the same cache keys as before I switched to Supabase (e.g. `posts/drafts`), which works fine for this situation even though the fetcher function doesn't use it, because it just acts as a cache-key that makes sense to humans reading the code like me.

And finally the fetcher function looks like this:

```javascript
export async function getDraftPosts() {
  const { data: posts, error } = await supabase
    .from('posts')
    .select('*')
    .eq('published', false)

  if (error) throw error.message
  return  posts
}
```

I haven't run into much weirdness yet and things are working fine. I will say that `auth.user()` and `auth.session()` just return the requested data rather than the `{ data, error }` returned by the others. So you can see I wrote a very thin fetcher function to get draft posts, destructure its return values, throw any errors and return the data. Whereas with `useSession` I just pass an anonymous arrow function as the fetcher: `useSWR('check_login', () => supabase.auth.session())` because `auth.session()` is _only_ returning the session object itself, or it's throwing an error; no errors-in-data and no destructuring needed.

Anyway I hope this helps a bit and good luck on your project!

**Reply by mhsnook** [reply]
Also, it's possible your original code will work if you remove the `async` and `await` in the fetcher function and just return the promise.

**Comment by psteinroe** [comment, marked as answer]
If you plan to use swr with supabase, be sure to check out [Supabase Cache Helpers](https://github.com/psteinroe/supabase-cache-helpers). It integrates supabase-js with swr and comes with mutation, subscription and convenience hooks for pagination as well as storage support. All logic is split up in reusable packages, hence adding support for `react-query` should be quite straight forward too.

```tsx
   // Define the query. Check out the other query hooks for pagination etc.
    const { data, count } = useQuery(
    client
        .from("contact")
        .select("id,username,ticket_number", { count: "exact" })
        .eq("username", "myusername"),
    "multiple",
    {
        revalidateOnFocus: false,
        revalidateOnReconnect: false,
    }
    );

    // Define the mutation. Update, Upsert, and Delete are also supported.
    const [insert] = useInsertMutation(client.from("contact"), "single");

    // Subscriptions are also supported.
    const { status } = useSubscription(
        client.channel('random'),
        {
            event: "*",
            table: "contact",
            schema: "public",
        },
        ["id"],
        { callback: (payload) => console.log(payload) }
    );

```

**Reply by foufrix** [reply]
It may be a stupid question, but how do you combine calls from an external API and Supabase one without having to add `SWR` next to `@supabase-cache-helpers/postgrest-swr` ?

My case is that I fetch data from supabase, but also has some NodeJS instances not hosted using supabase that give me other values unrelated to my database.

Also, does `postgres-swr` support https://github.com/koba04/swr-devtools ?

**Reply by foufrix** [reply]
@psteinroe

**Reply by psteinroe** [reply]
hey @foufrix. why are you so opposed to using swr directly? cache helpers is just a wrapper around swr. its literally just a hook that calls `useSWR`. So yes, dev tools work because you are just using plain old swr, even with cache helpers. this is also why 
>  without having to add SWR next to @supabase-cache-helpers/postgrest-swr

does not make sense, because you are using SWR anyways if you use cache helpers. I also have similar cases, and just use swr directly to call the custom api.
